Allow to select Qt's default style

sledgehammer999 opened the pull request:
Relevant prior PR #21553

Fusion is still the default.
Here is how it looks on a default profile:
![Untitled](https://github.com/user-attachments/assets/c4323e33-ee1d-46e8-a769-76b9059bafa8)

Pentaphon commented:
How is "system" different from the other styles?

sledgehammer999 commented:
>How is "system" different from the other styles?

The tooltip says it: "Let Qt decide the QStyle for this system" -> Aka whatever Qt decides as default.

Pentaphon commented:
> The tooltip says it: "Let Qt decide the QStyle for this system" -> Aka whatever Qt decides as default, which will be one of the other named styles.

Ok so assuming each of the combobox options gets a tooltip explaining what each style can do, what else is left to do for this PR?

sledgehammer999 commented:
>Ok so assuming each of the combobox options gets a tooltip explaining what each style can do

No. Only the "system" one, because it is a special value inserted by us. The other styles come from a dynamic list offered by Qt. Qt doesn't guarantee neither the name of the styles nor what each does. But unofficially the style names seem to be stable for several years now.

>what else is left to do for this PR?

From my end, nothing. I wait review.

glassez commented on src/gui/optionsdialog.cpp:
```diff
@@ -1690,20 +1697,34 @@ void OptionsDialog::initializeStyleCombo()
+    if (selectedStyleName.compare(u"system"_s, Qt::CaseInsensitive) == 0)
     {
-        if (selectedStyleName.compare(styleNames.at(i), Qt::CaseInsensitive) == 0)
+        m_ui->comboStyle->addItem(tr("System", "System default QtStyle"), u"system"_s);
```
```suggestion
        m_ui->comboStyle->addItem(tr("System", "System default Qt style"), u"system"_s);
```

glassez commented on src/gui/optionsdialog.cpp:
```diff
@@ -1690,20 +1697,34 @@ void OptionsDialog::initializeStyleCombo()
     {
-        if (selectedStyleName.compare(styleNames.at(i), Qt::CaseInsensitive) == 0)
+        m_ui->comboStyle->addItem(tr("System", "System default QtStyle"), u"system"_s);
+        m_ui->comboStyle->setItemData(0, tr("Let Qt decide the QStyle for this system"), Qt::ToolTipRole);
```
I would avoid internal names (e.g. `QStyle`) from being used in UI.

```suggestion
        m_ui->comboStyle->setItemData(0, tr("Let Qt decide the style for this system"), Qt::ToolTipRole);
```

glassez reviewed (commented):
It conflicts with #21599 so I would wait for #21599 is merged and conflicts are resolved before final review.

Chocobo1 commented on src/gui/optionsdialog.cpp:
```diff
@@ -454,7 +458,10 @@ void OptionsDialog::saveBehaviorTabOptions() const
 
 #ifdef Q_OS_WIN
-    pref->setStyle(m_ui->comboStyle->currentText());
+    if (const auto sysVariant = m_ui->comboStyle->currentData(); sysVariant.isValid())
```
```suggestion
    if (const QVariant sysVariant = m_ui->comboStyle->currentData(); sysVariant.isValid())
```

Chocobo1 commented on src/gui/optionsdialog.cpp:
```diff
@@ -1690,20 +1697,34 @@ void OptionsDialog::initializeStyleCombo()
+    if (selectedStyleName.compare(u"system"_s, Qt::CaseInsensitive) == 0)
     {
-        if (selectedStyleName.compare(styleNames.at(i), Qt::CaseInsensitive) == 0)
+        m_ui->comboStyle->addItem(tr("System", "System default Qt style"), u"system"_s);
```
`System` -> `System default` or `Qt default`.
How about those suggestions?

sledgehammer999 replied:
IMO, if we add "default" to any entry it will confuse users as to what we recommend.

sledgehammer999 replied:
>if I don't know in advance what it means.

Honest question: Do you think the tooltip won't help with that?

glassez replied:
> > if I don't know in advance what it means.
> 
> Honest question: Do you think the tooltip won't help with that?

Maybe. Unless user initially gets a false impression about its meaning and chooses it without looking at tooltip.

glassez replied:
Anyway, let's not dwell on it. It can be renamed later based on real user feedbacks.

glassez commented on src/gui/optionsdialog.cpp:
```diff
@@ -244,6 +244,10 @@ void OptionsDialog::loadBehaviorTabOptions()
     initializeStyleCombo();
 
+#ifdef Q_OS_WIN
+    m_ui->labelStyleHint->setText(tr("%1 is recommended for best compatibility with Windows dark mode", "Fusion is recommended for best compatibility with Windows dark mode").arg(u"Fusion"_s));
```
```suggestion
    m_ui->labelStyleHint->setText(tr("%1 is recommended for best compatibility with Windows dark mode"
            , "Fusion is recommended for best compatibility with Windows dark mode").arg(u"Fusion"_s));
```

glassez commented on src/gui/optionsdialog.cpp:
```diff
@@ -244,6 +244,10 @@ void OptionsDialog::loadBehaviorTabOptions()
 
     initializeStyleCombo();
 
+#ifdef Q_OS_WIN
+    m_ui->labelStyleHint->setText(tr("%1 is recommended for best compatibility with Windows dark mode", "Fusion is recommended for best compatibility with Windows dark mode").arg(u"Fusion"_s));
+#endif
```
I would move it in `initializeStyleCombo()` and rename it to `initializeStyleOptions()`

glassez reviewed (commented):
Are so many changes of `optionsdialog.ui` really required?

glassez commented on src/gui/optionsdialog.cpp:
```diff
@@ -1689,17 +1696,24 @@ bool OptionsDialog::isSplashScreenDisabled() const
 void OptionsDialog::initializeStyleCombo()
 {
 #ifdef Q_OS_WIN
+    m_ui->comboStyle->addItem(tr("System", "System default Qt style"), u"system"_s);
```
So is "System" item out of the sort order? Waiting for @Chocobo1 claim...

sledgehammer999 replied:
Having it integrated into the sorted items will probably require non-clean code.
Also, IMO, it makes sense for "system" (in some sense a "default") to always be first in the list. We can drive home this intention with the usage of https://doc.qt.io/qt-6/qcombobox.html#insertSeparator
What do you think?

glassez replied:
> What do you think?

IMO, it is good idea to separate special item.


> Having it integrated into the sorted items will probably require non-clean code.

This should not cause problems if all the items were handled uniformly, i.e. they would have a separate display name and value stored in combobox, as you did for the System style. BTW, it would also allow to provide consistently formatted display names, i.e. "Windows", "WindowsVista", "Windows11", "Fusion", "System".

Chocobo1 replied:
>We can drive home this intention with the usage of https://doc.qt.io/qt-6/qcombobox.html#insertSeparator
What do you think?

I'm ok with it.

sledgehammer999 commented:
>Are so many changes of optionsdialog.ui really required?

To be honest this is Qt Designer output. I just added the label and tweaked a few settings on that label. I left the other things unchanged but Qt Designer decides to rearranged things when saving.
If you want, I can try to interactively add only the relevant changes with `git add -p optionsdialog.ui`

glassez commented:
> I only added the label and tweaked a few settings on that label. I left the other things unchanged but Qt Designer decides to rearranged things when saving.

We usually insist that only relevant changes are added in `.ui` files, manually or in some other way that avoids unnecessary changes.

sledgehammer999 commented:
- I added the separator and updated the pic in the first post
- I am not sure how to indent the split line. I am using VSCode now and I don't have QtCreator's automatic indentation of split lines.
- I am not sure how to further filter the changes in `optionsdialog.ui` without potentially breaking it.

glassez commented:
> * I am not sure how to further filter the changes in `optionsdialog.ui` without potentially breaking it.

Dealing with such changes are really messy. But UI file is just XML. It's not so hard to edit it manually.
If I need to add some kind of control to some widget, I usually do the following. I add it in the Designer, then I find a portion of the XML code related to it, copy it, reset the changes in .ui and paste copied code manually (perhaps also adjusting the properties of adjacent elements, for example, row/column numbers). Yes, it is more difficult than just using a Designer, but it produces a readable "diff" and our history does not contain these useless/confusing changes.

glassez commented:
> > * I am not sure how to further filter the changes in `optionsdialog.ui` without potentially breaking it.
> 
> Dealing with such changes are really messy. But UI file is just XML. It's not so hard to edit it manually. If I need to add some kind of control to some widget, I usually do the following. I add it in the Designer, then I find a portion of the XML code related to it, copy it, reset the changes in .ui and paste copied code manually (perhaps also adjusting the properties of adjacent elements, for example, row/column numbers). Yes, it is more difficult than just using a Designer, but it produces a readable "diff" and our history does not contain these useless/confusing changes.

I believe all you need is to replace the following fragment (of original `optionsdialog.ui`):
```xml
<item row="2" column="2">
 <spacer name="spacerStyle">
  <property name="orientation">
   <enum>Qt::Orientation::Horizontal</enum>
  </property>
  <property name="sizeHint" stdset="0">
   <size>
    <width>40</width>
    <height>20</height>
   </size>
  </property>
 </spacer>
</item>
```
...with the following:
```xml
<item row="2" column="2">
 <widget class="QLabel" name="labelStyleHint"/>
</item>
```

oopwow commented:
> Соответствующий предыдущий PR [# 21553](https://github.com/qbittorrent/qBittorrent/pull/21553)
> 
> Fusion по-прежнему используется по умолчанию. Вот как он выглядит в профиле по умолчанию: ![Без названия](https://private-user-images.githubusercontent.com/273315/376405572-50bb80d5-af11-447e-a9f4-ada391e9bd9b.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkwOTA1NDYsIm5iZiI6MTcyOTA5MDI0NiwicGF0aCI6Ii8yNzMzMTUvMzc2NDA1NTcyLTUwYmI4MGQ1LWFmMTEtNDQ3ZS1hOWY0LWFkYTM5MWU5YmQ5Yi5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxNlQxNDUwNDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MWE0YjQxZDE3NDI1NmRlODEwOTBlM2U3OTExZDc5ZWZlZWU5YjUzMjI4MDBmOGE2NzFjYTA1MTYyYzZlYmRhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.qdoMw4rkchpuIvSH3QR7A450sHYG22P6Gi-ScqdFOLw)

Hello. In my opinion, theme names should be clear. The name "Windows" is not suitable, since "qBittorrent" exists not only for the Windows OS.

Theme names that are suitable for all OS:
"Light" - _white theme_
"Dark" - _black theme_
"Sync (theme color with computer OS)" - _fusion theme_
or
"Fusion (theme color with computer OS)"




P.S. In my opinion, by default, the "Light" theme should be enabled (as in Windows OS). Windows OS has a million-strong audience, they are according to user, they chose the best default color.

As a last resort, the choice of theme color can be added during installation, here:
![373903255-490fbd75-1b68-468e-a8f5-18fb801b6017](https://github.com/user-attachments/assets/b583ce7c-609a-41ac-854c-211cd20396ea)



_Please try to make the colors of the dark theme and light theme pleasing to the eye, as well as their icons and everything else. So that qBittorrent looks like a modern application.

Examples of dark and light theme color shades:_ https://github.com/qbittorrent/qBittorrent/issues/21521

glassez commented:
@oopwow 
This topic isn't about color schemes.

qBittUser commented:
@oopwow
> The name "Windows" is not suitable, since "qBittorrent" exists not only for the Windows OS.

There's no optional Qt style setting on different OS in this PR and qBittorrent relies on what Qt has to offer, so naming, available settings and OS compatibility is kept to be as close as possible to whatever Qt offers and things might change later on.

Including GUI or webUI changes in Windows NSIS installer instead of only built-in options afterwards is not worth it, I haven't even seen any software to allow to choose theme in the setup file.

* Examples of dark and light theme color shades_ #21521

Your multiple issue/idea/feature request in a single issue report needs to closed and be marked as a duplicate of many others or at least split up as totally separate discussions.

> Please try to make the colors ...

This Pull Request is to at first add extra basic options. There's already other pull requests that gives even more options. Too many changes at the same pull request would make it hard for anyone to quickly do a proper test and review.

Default theme icon color change should be discussed in a more appropriate place, not delay or confuse others by posting off-topic extra ideas here or post obvious duplicate ideas that has been already shared elsewhere by you or others.

If you didn't know, then You can already modify many default colors and icons since v4.6 was released or just use custom theme files.

sledgehammer999 commented:
I have addressed the last comments and posted a new screenshot.
